r1cs/lib/r1cs/Ops.cpp

```diff
@@ -86,8 +106,53 @@ mlir::Block *CircuitDefOp::addEntryBlock() {
   return block;
 }
 
-void CircuitDefOp::build(OpBuilder &builder, OperationState &state, llvm::StringRef name) {
-  state.addAttribute(SymbolTable::getSymbolAttrName(), builder.getStringAttr(name));
-  state.addRegion();
+LogicalResult CircuitDefOp::verify() {
+
+  // === Step 1: Check that each arg_attrs key is a valid argument index ===
+  if (getArgAttrs().has_value()) {
+    DictionaryAttr dict = *getArgAttrs();
+    unsigned numArgs = getBody().front().getNumArguments();
+
+    for (auto attr : dict) {
+      unsigned index = 0;
+      if (attr.getName().strref().getAsInteger(10, index)) {
+        return emitOpError() << "invalid key '" << attr.getName()
+                             << "' in 'arg_attrs': expected integer index";
+      }
+      if (index >= numArgs) {
+        return emitOpError() << "argument index " << index << " out of bounds (only " << numArgs
+                             << " arguments)";
+      }
+      if (!attr.getValue().isa<r1cs::PublicAttr>()) {
+        return emitOpError() << "invalid attribute for argument " << index
+                             << ": expected #r1cs.pub";
```

Comment:
Can you get `r1cs.pub` via `PublicAttr::name()` or something? Prefer this to allow for possible future renaming.